### ISSUE TITLE
dead_code lint wrongly warns about "never used" functions that are, in fact used

### ISSUE THREAD

**Issue by glandium** [rust-lang/rust#126289]
### Code

```rust
mod ffi {
    use super::*;

    extern "C" {
        pub fn DomPromise_AddRef(promise: *const Promise);
        pub fn DomPromise_Release(promise: *const Promise);
    }
}

#[repr(C)]
#[allow(unused)]
pub struct Promise {
    private: [u8; 0],
    __nosync: ::std::marker::PhantomData<::std::rc::Rc<u8>>,
}

pub unsafe trait RefCounted {
    unsafe fn addref(&self);
    unsafe fn release(&self);
}

unsafe impl RefCounted for Promise {
    unsafe fn addref(&self) {
        ffi::DomPromise_AddRef(self)
    }
    unsafe fn release(&self) {
        ffi::DomPromise_Release(self)
    }
}
```

### Output with 1.79.0:

nothing

### Output with 1.80.0:

```
warning: function `DomPromise_AddRef` is never used
 --> src/lib.rs:5:16
  |
5 |         pub fn DomPromise_AddRef(promise: *const Promise);
  |                ^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: function `DomPromise_Release` is never used
 --> src/lib.rs:6:16
  |
6 |         pub fn DomPromise_Release(promise: *const Promise);
  |                ^^^^^^^^^^^^^^^^^^

warning: `playground` (lib) generated 2 warnings
```

The original code didn't have the `#[allow(unused)]`, which was added because of the "never constructed" dead_code lint, which I guess would be #126169. Adding a constructor does make the function never used errors go away, so it seems `#[allow(unused)]` doesn't have enough power.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23126289.20dead_code.20lint.20wrongly.20warns.20about.20.22never.20used.22.20fun.E2.80.A6)).

@rustbot label -I-prioritize +P-medium

**Comment by apiraino** [comment]
bisected to https://github.com/rust-lang-ci/rust/commit/c13710982e6438501456f5451060b45b80d38530

</details>

cc @mu001999

**Comment by glandium** [comment]
#126169 is about the struct itself, and there's a allow(unused). This is about the side effects.
